internal/controller/nnf_systemconfiguration_controller.go

```diff
@@ -190,12 +205,130 @@ func (r *NnfSystemConfigurationReconciler) createNamespaces(ctx context.Context,
 	return nil
 }
 
+// labelsAndTaints will ensure that all rabbits are labeled and it will
+// apply appropriate taints to the rabbits.
+//
+// The first time through it will ensure that each rabbit is labeled and that
+// it has the NoSchedule taint.
+// The second time through it will add the NoExecute taint.
+// The third time through it will remove the NoExecute taint.
+//
+// This does not apply the NoExecute taint until we're certain that every
+// rabbit already has the NoSchedule taint, so we don't play musical chairs
+// with the pods that we're evicting, chasing them from rabbit to rabbit.
+// A breadcrumb (a label) is left on the node to remind us when we've already
+// done the add/remove of the NoExecute taint, so we don't repeat that cycle.
+func (r *NnfSystemConfigurationReconciler) labelsAndTaints(ctx context.Context, validRabbits map[string]struct{}) (bool, error) {
+	var err error
+
+	// Pass 1: apply cray.nnf.node label and NoSchedule taint.
+	// Pass 2: apply NoExecute taint.
+	// Pass 3: apply cray.nnf.node.cleared label and clear the NoExecute taint.
+	var clearNoExecute corev1.TaintEffect = "clear-no-execute"
+	taintEffectPerPass := []corev1.TaintEffect{
+		corev1.TaintEffectNoSchedule,
+		corev1.TaintEffectNoExecute,
+		clearNoExecute,
+	}
+
+	updatedNode := false
+	for pass, effect := range taintEffectPerPass {
+
+		if updatedNode {
+			// The previous pass must be a no-op before we can run the next pass.
+			break
+		}
+
+		for name := range validRabbits {
+			doUpdate := false
+			node := &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: name,
+				},
+			}
+			log := r.Log.WithValues("Node", client.ObjectKeyFromObject(node))
+			if err = r.Get(ctx, client.ObjectKeyFromObject(node), node); err != nil {
+				// Maybe it's been removed for administrative purposes and the
+				// SystemConfiguration hasn't been updated.
+				log.Info("unable to find node", "error", err)
+				continue
+			}
+			labels := node.GetLabels()
+			if labels == nil {
+				labels = make(map[string]string)
+			}
+
+			if pass > 0 {
+				// On this node, do later passes only if we haven't already
+				// completed all passes here.
+				if _, present := labels["cray.nnf.node.cleared"]; present {
```

Comment:
Looks like this label sticks around for the case when new nodes are added. I'm not sure I'll remember what `cleared` means. Maybe `taintsAndLabelsComplete`?